File: Leetcode 242 - valid-anagram/242.valid-anagram.cpp
/*
 * @lc app=leetcode id=242 lang=cpp
 *
 * [242] Valid Anagram
 */

// @lc code=start
#include <iostream>
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
  bool isAnagram(string s, string t) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    unordered_map<char, int> countS;

    if (s.size() != t.size()) {
      return false;
    }

    for (int i = 0; i < s.size(); i++) {
      countS[s[i]]++;
      countS[t[i]]--;
    }

    for (const auto &pair : countS) {
      if (pair.second != 0) {
        return false;
      }
    }

    return true;
  }
};

int main() {
  Solution sol;

  cout << boolalpha;
  cout << sol.isAnagram("anagram", "nagaram") << "\n"; // true
  cout << sol.isAnagram("rat", "car");                 // false
}
// @lc code=end

File: Leetcode 242 - valid-anagram/242.valid-anagram.java
/*
 * @lc app=leetcode id=242 lang=java
 *
 * [242] Valid Anagram
 */

// @lc code=start

class Solution {

  public boolean isAnagram(String s, String t) {
    int[] countS = new int[26];
    for (char c : s.toCharArray()) {
      countS[((int) c) - 97] = countS[((int) c) - 97] + 1;
    }
    int[] countT = new int[26];
    for (char c : t.toCharArray()) {
      countT[((int) c) - 97] = countT[((int) c) - 97] + 1;
    }
    for (int i = 0; i < 26; i++) {
      if (countS[i] != countT[i]) {
        return false;
      }
    }
    return true;
  }

  public static void main(String[] args) {
    Solution solution = new Solution();

    System.out.println(solution.isAnagram("anagram", "nagaram")); // true
    System.out.println(solution.isAnagram("rat", "car")); // false
  }
}
// @lc code=end
